extensions/ql-vscode/src/databases/github-database-prompt.ts

```diff
@@ -0,0 +1,137 @@
+import { window } from "vscode";
+import { RestEndpointMethodTypes } from "@octokit/plugin-rest-endpoint-methods";
+import { Octokit } from "@octokit/rest";
+import { showNeverAskAgainDialog } from "../common/vscode/dialog";
+import { getLanguageDisplayName } from "../common/query-language";
+import {
+  downloadGitHubDatabaseFromUrl,
+  promptForLanguage,
+} from "./database-fetcher";
+import { withProgress } from "../common/vscode/progress";
+import { DatabaseManager } from "./local-databases";
+import { CodeQLCliServer } from "../codeql-cli/cli";
+import { AppCommandManager } from "../common/commands";
+import { GitHubDatabaseConfig } from "../config";
+
+export type CodeqlDatabase =
+  RestEndpointMethodTypes["codeScanning"]["listCodeqlDatabases"]["response"]["data"][number];
+
+export async function findGitHubDatabasesForRepository(
+  octokit: Octokit,
+  owner: string,
+  repo: string,
+): Promise<CodeqlDatabase[]> {
+  const response = await octokit.rest.codeScanning.listCodeqlDatabases({
+    owner,
+    repo,
+  });
+
+  return response.data;
+}
+
+/**
+ * Prompt the user to download a database from GitHub. This is a blocking method, so this should
+ * almost never be called with `await`.
+ */
+export async function promptGitHubDatabaseDownload(
+  octokit: Octokit,
+  owner: string,
+  repo: string,
+  databases: CodeqlDatabase[],
+  config: GitHubDatabaseConfig,
+  databaseManager: DatabaseManager,
+  storagePath: string,
+  cliServer: CodeQLCliServer,
+  commandManager: AppCommandManager,
+): Promise<void> {
+  const languages = databases.map((database) => database.language);
+
+  const databasesMessage =
+    databases.length === 1
+      ? `This repository has an origin (GitHub) that has a ${getLanguageDisplayName(
+          languages[0],
+        )} CodeQL database.`
+      : `This repository has an origin (GitHub) that has ${joinLanguages(
+          languages,
+        )} CodeQL databases.`;
+
+  const connectMessage =
+    databases.length === 1
+      ? `Connect to GitHub and download the existing database?`
+      : `Connect to GitHub and download any existing databases?`;
+
+  const answer = await showNeverAskAgainDialog(
+    `${databasesMessage} ${connectMessage}`,
```

Comment:
Any specific reason why this string is split in two? It seems that the condition is the same in both cases?

Reply:
Good point, I've merged these two messages.